README.md
[![GH release](https://img.shields.io/github/v/release/shaarli/netscape-bookmark-parser)](https://github.com/shaarli/netscape-bookmark-parser/releases)
![GH license](https://img.shields.io/github/license/shaarli/netscape-bookmark-parser)
![GH stars](https://img.shields.io/github/stars/shaarli/netscape-bookmark-parser)
![Build Status](https://github.com/shaarli/netscape-bookmark-parser/workflows/PHP%20CI/badge.svg)

Shaarli Netscape Bookmark Parser
================================

This library provides a decoder that is able of parsing Netscape bookmarks (as exported by common Web browsers and bookmarking services), and an encoder that is able to export data to bookmarks format.

🚀 Installation
---------------

Using [Composer](https://getcomposer.org/) ([package](https://packagist.org/packages/shaarli/netscape-bookmark-parser)):

```bash
$ composer require shaarli/netscape-bookmark-parser
```

🔥 Usage
--------

### Import from Bookmark file

```php
<?php
// parse.php
require_once 'vendor/autoload.php';

use Shaarli\NetscapeBookmarkParser\NetscapeBookmarkParser;

$parser = new NetscapeBookmarkParser();
$bookmarks = $parser->parseFile('./tests/Fixtures/Encoder/input/netscape_basic.htm');
var_dump($bookmarks);
```

Above example will return following array:

```
array(2) {
  [0] => array(7) {
    ["name"] => string(12) "Secret stuff"
    ["image"] => NULL
    ["url"] => string(19) "https://private.tld"
    ["tags"] => array(2) {
      [0] => string(7) "private"
      [1] => string(6) "secret"
    }
    ["description"] => string(52) "Super-secret stuff you're not supposed to know about"
    ["dateCreated"] => int(971175336)
    ["public"] => bool(false)
  }
  [1] => array(7) {
    ["name"] => string(12) "Public stuff"
    ["image"] => NULL
    ["url"] => string(17) "http://public.tld"
    ["tags"] => array(3) {
      [0] => string(6) "public"
      [1] => string(5) "hello"
      [2] => string(5) "world"
    }
    ["description"] => NULL
    ["dateCreated"] => int(1456433748)
    ["public"] => bool(true)
  }
}
```

### Export to file

```php
<?php
// export.php
require_once 'vendor/autoload.php';

use Shaarli\NetscapeBookmarkParser\NetscapeBookmarkParser;

$data = [
    [
        'name'        => 'shaarli/Shaarli',
        'url'         => 'https://github.com/shaarli/Shaarli',
        'image'       => 'data:image/png;base64, ...'
        'description' => 'The personal, minimalist, super-fast, database free, bookmarking service',
        'public'      => true,
        'tags'        => ['shaarli'],
    ],
    [
        'name'         => 'shaarli/netscape-bookmark-parser',
        'url'          => 'https://github.com/shaarli/netscape-bookmark-parser',
        'public'       => false,
        'dateCreated'  => '1612572000',
        'dateModified' => '1612572000',
        'tags'         => ['shaarli'],
    ],
];

$parser = new NetscapeBookmarkParser();

$parser->export($data, './export.htm');
```

Above example will write following string into file:

```xml
<!DOCTYPE NETSCAPE-Bookmark-file-1>
<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=UTF-8">
<TITLE>Bookmarks</TITLE>
<H1>Bookmarks</H1>
<DL><p>
<DT><H3>shaarli</H3>
<DL><p>
<DT><A HREF="https://github.com/shaarli/Shaarli" PRIVATE="0" ICON="data:image/png;base64, ...">shaarli/Shaarli</A>
<DD>The personal, minimalist, super-fast, database free, bookmarking service
<DT><A HREF="https://github.com/shaarli/netscape-bookmark-parser"> ADD_DATE="1612572000" LAST_MODIFIED="1612572000">shaarli/netscape-bookmark-parser</A>
<\DL><p>
<\DL><p>
```

💻 Dependencies
---------------

**Shaarli Netscape Bookmark Parser** requires the following dependencies:

- PHP
- Composer
- Make (optional)

---

### 🐘 PHP

Learn how to install PHP from official documentation [here](https://www.php.net/manual/en/install.php)

---

### 📦 Composer

Learn how to install Composer from official documentation [here](https://getcomposer.org/download/)

---

### 🛠 Make

#### 🐧 Install Make (Linux)

On linux machine enter following command

```bash
$ sudo apt-get install --assume-yes make
```

#### 🏁 Install Make (Windows)

On windows machine you will need to install [cygwin](http://www.cygwin.com/) or [GnuWin make](http://gnuwin32.sourceforge.net/packages/make.htm) first to execute make script.

#### 🍎 Install Make (OSX)

Make should be available by default on OSX system, but you can upgrade make version with following command

```bash
$ brew install make
```

---

📑 About
--------

### ⚡ A word from Kafene

The motivations behind developing this parser are the following:
- the [Netscape format](https://msdn.microsoft.com/en-us/library/aa753582%28v=vs.85%29.aspx) has a very loose specification: no [DTD](https://en.wikipedia.org/wiki/Document_type_definition) nor [XSL stylesheet](https://en.wikipedia.org/wiki/XSL) to constrain how data is formatted.
- software and web services export bookmarks using a wild variety of attribute names and values.
- using standard SAX or DOM parsers is thus not straightforward.

How it works:
- the input bookmark file is trimmed and sanitized to improve parsing results
- the resulting data is then parsed using [PCRE](http://www.pcre.org/) patterns to match attributes and values corresponding to the most likely:
    - attribute names: `description` vs. `note`, `tags` vs. `labels`, `date` vs. `time`, etc.
    - data formats: `comma,separated,tags` vs. `space separated labels`, UNIX epochs vs. human-readable dates, newlines & carriage returns, etc.
- an associative array containing all successfully parsed links with their attributes is returned

### ⚡ Shaarli community fork

This friendly fork is maintained by the Shaarli community at https://github.com/shaarli/netscape-bookmark-parser and is used by the open-source [Shaarli](https://github.com/shaarli/Shaarli) bookmarking service. This is a community fork of the original [netscape-bookmark-parser](https://github.com/kafene/netscape-bookmark-parser) project by [Kafene](http://kafene.org/).

🤝 Contributing
---------------

Thank you for your interrest in contributing to **Shaarli Netscape Bookmark Parser**.

Please review the [code of conduct](./CODE_OF_CONDUCT.md) and [contribution guidelines](./CONTRIBUTING.md) before starting to work on any features.

If you want to open an issue, please check first if it was not [reported already](https://github.com/shaarli/netscape-bookmark-parser/issues) before creating a new one.

📜 License
----------

Copyrights (c) 2021 &quot;Shaarli community&quot;

[![License](https://img.shields.io/badge/Licence-MIT-green.svg)](LICENSE)
Distributed under the MIT license.

If you like **Shaarli Netscape Bookmark Parser** please star or tweet about it:

[![GitHub stars](https://img.shields.io/github/stars/shaarli/netscape-bookmark-parser?style=social)](https://github.com/shaarli/netscape-bookmark-parser/stargazers)
[![Twitter](https://img.shields.io/twitter/url?style=social&url=https%3A%2F%2Fgithub.com%2Fshaarli%2Fnetscape-bookmark-parser)](https://twitter.com/intent/tweet?text=Wow:&url=https%3A%2F%2Fgithub.com%2Fshaarli%2Fnetscape-bookmark-parser)

🙏 Acknowledgements
-------------------

- [Shaarli Community](https://github.com/shaarli)
- [Kafene](https://github.com/kafene)
- [VirtualTam](https://github.com/virtualtam)
- [TangoMan75](https://github.com/TangoMan75)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

**Please read the changelog carefully, this is a major release, there are breaking changes.**

### Added
- `NetscapeBookmarkEncoder` and `NetscapeBookmark.Decoder` classes.
- CHANGELOG.md and contributor CODE_OF_CONDUCT.md.
- github ISSUE_TEMPLATE.md for bug reports and feature requests.
- `supportsDecoding` method and `self::FORMAT` class constant (for future Interface conformity and Symfony services integration).
- `BookmarksDataFormatter` class to use before export to have bookmarks ordered by tag.
- `BookmarksDataConverter` and `LegacyBookmarksDataConverter` class to convert old bookmarks to new data format.
- PHPUnit configuration file: phpunit.xml.dist

### Changed
- Remove `Katzgrau-KLogger` dependency.
- `parseBoolean` method should always return a boolean value.
- Renamed properties to be compatible with "https://schema.org" structured data vocabulary:
```php
[
    'title' => 'name',
    'icon'  => 'image',
    'uri'   => 'url',
    'tags'  => 'tags',
    'note'  => 'description',
    'time'  => 'dateCreated',
    'pub'   => 'public'
];
```
- Constructor: Replace parameters from constructor with `$context` array and class constants.
- Updated PHPUnit to the latest version (9.5).
- Undefined values should remain `null`
	- Decoder should not set default `$item['name'] = 'untitled'`.
	- Decoder should not set `$item['dateCreated'] = time()` as default.
- Removed default Tags ; This is not the decoder job to set default values.
- Removed default public/private status ; This is not the decoder job to set default public status.
- Moved source files and updated file structure.
- Travis CI replaced with GitHub actions.
- Updated PHPCS linter configuration.
- Updated README.md
- Improved `Makefile`.

### Fixed
- TRUE_PATTERN yes appears twice; ordered alphabetically.

### Removed
- Drop support for PHP < 7.3. PHP 7.1 and 7.2 have reached end of life.
- travis.yml, .editorconfig, .gitattributes
- `NetscapeBookmarkParser` should be removed next major release.

## [v3.2.0](https://github.com/shaarli/netscape-bookmark-parser/releases/tag/v3.2.0) - 2021-02-22

### Added
- Parsing 'ICON' attribute.

### Fixed
- Improve PHPCS linter configuration

## [v3.1.0](https://github.com/shaarli/netscape-bookmark-parser/releases/tag/v3.1.0) - 2021-01-25

### Changed
- Scale down 'items' variable scope to method level. Class is now stateless.
- Refactor 'parseString' method
- Replaced idle 'sanitized' variable in 'sanitizeString' method

## [v3.0.2](https://github.com/shaarli/netscape-bookmark-parser/releases/tag/v3.0.2) - 2021-01-17

### Fixed
- Remove unknown parameter in `parseBoolean` function calls

## [v3.0.1](https://github.com/shaarli/netscape-bookmark-parser/releases/tag/v3.0.1) - 2020-11-03

### Added
- README: installation section (composer)

### Fixed
- README: fix default example with v3.0.0 values

## [v3.0.0](https://github.com/shaarli/netscape-bookmark-parser/releases/tag/v3.0.0) - 2020-11-03

**Please read the changelog carefully, this is a major release, there are breaking changes.**

### Added
- Parsing now supports tags with multiple words separated by whitespaces.
- Official PHP compatibility with PHP 7.3, 7.4 and 8.0.
- Follow and enforce PSR-12 coding style

### Changed
- **Bookmark's tags now returns an array of tags instead of tags string separated by whitespaces.**

### Removed
- Drop support for PHP 5.6 and 7.0.

## [v2.2.0](https://github.com/shaarli/netscape-bookmark-parser/releases/tag/v2.2.0) - 2020-06-06
### Added
- Support Shaarli extended format: allow <a> tag in bookmark description

### Changed
- Log to current working directory by default instead of library path
- Do not replace tab character to preserve bookmarks formatting

### Fixed
- Error while trying to import a file without bookmarks
- Undefined time variable

## [v2.1.0](https://github.com/shaarli/netscape-bookmark-parser/releases/tag/v2.1.0) - 2018-10-06
### Added
- Add PHP 7.1 and 7.2 to the Travis test environments
- Add support for microsecond timestamps

### Changed
- Update test dependencies
- Update regexes to PCRE2 for PHP 7.3

### Removed
- Drop support for PHP 5.3.x, 5.4.x and 5.5.x

## [v2.0.5](https://github.com/shaarli/netscape-bookmark-parser/releases/tag/v2.0.5) - 2018-01-30
## Fixed
- Fix scuttle description on multiple lines
- Improve sanitizing regexp, preventing trimming actual content

## [v2.0.4](https://github.com/shaarli/netscape-bookmark-parser/releases/tag/v2.0.4) - 2017-07-30
### Changed
- Pin Travis builds to `precise` to ensure PHP 5.3 compatibility

## [v2.0.3](https://github.com/shaarli/netscape-bookmark-parser/releases/tag/v2.0.3) - 2017-07-30
### Changed
- Update nested folder name parsing to sanitize resulting tags

## [v2.0.2](https://github.com/shaarli/netscape-bookmark-parser/releases/tag/v2.0.2) - 2017-06-10
### Changed
- Update note/description parsing to support Scuttle exports

## [v2.0.1](https://github.com/shaarli/netscape-bookmark-parser/releases/tag/v2.0.1) - 2017-03-08
### Changed
- Allow specifying the log directory

## [v2.0.0](https://github.com/shaarli/netscape-bookmark-parser/releases/tag/v2.0.0) - 2017-02-19
### Added
- Log import into a log file

### Changed
- NetscapeBookmarkParser class is now under `Shaarli` namespace.

## [v1.1.1](https://github.com/shaarli/netscape-bookmark-parser/releases/tag/v1.1.1) - 2017-02-15
### Added
- Add `CHANGELOG.md`
- Enable Composer cache for Travis builds

### Fixed
- Keep Composer metadata in Git-generated archives

## [v1.0.1](https://github.com/shaarli/netscape-bookmark-parser/releases/tag/v1.0.1) - 2016-08-09
### Changed
- Mark this repository as being a friendly fork maintained by the Shaarli community
- Add/update CI badges

### Fixed

src/Encoder/NetscapeBookmarkDecoder.php
<?php

declare(strict_types=1);

namespace Shaarli\NetscapeBookmarkParser\Encoder;

use Psr\Log\LoggerInterface;
use Psr\Log\NullLogger;

/**
 * Generic Netscape bookmark decoder.
 */
class NetscapeBookmarkDecoder
{
    public const FORMAT = 'netscape';

    /**
     * keep_nested_tags: boolean
     *     Tag links with parent folder names
     */
    public const KEEP_NESTED_TAGS = 'keep_nested_tags';

    /**
     * normalize_dates: boolean
     *     Whether parsed dates are expected to fall within a given date/time interval
     */
    public const NORMALIZE_DATES = 'normalize_dates';

    /**
     * date_range: string
     *     Delta used to compute the "acceptable" date/time interval
     */
    public const DATE_RANGE = 'date_range';

    public const DEFAULT_KEEP_NESTED_TAGS = true;
    public const DEFAULT_NORMALIZE_DATES = true;
    public const DEFAULT_DATE_RANGE = '30 years';

    public const TRUE_PATTERN = '1|\+|array|checked|ok|okay|on|one|t|true|y|yes';
    public const FALSE_PATTERN = '-|0|die|empty|exit|f|false|n|neg|nil|no|null|off|void|zero';

    private $defaultContext = [
        self::KEEP_NESTED_TAGS => self::DEFAULT_KEEP_NESTED_TAGS,
        self::NORMALIZE_DATES  => self::DEFAULT_NORMALIZE_DATES,
        self::DATE_RANGE       => self::DEFAULT_DATE_RANGE,
    ];

    /**
     * @var LoggerInterface
     */
    private $logger;

    public function __construct(array $defaultContext = [], LoggerInterface $logger = null)
    {
        $this->defaultContext = array_merge($this->defaultContext, $defaultContext);
        $this->logger = $logger ?? new NullLogger();
    }

    /**
     * Checks whether the deserializer can decode from given format.
     *
     * @param string $format Format name
     *
     * @return bool
     */
    public function supportsDecoding(string $format): bool
    {
        return self::FORMAT === $format;
    }

    /**
     * Decodes a string containing Netscape-formatted bookmarks into PHP data.
     *
     * Output format:
     *
     * array(2) {
     *     [0]=> array(7) {
     *         ['name'] => string(9) 'Some page'
     *         ['image'] => string(26) 'data:image/png;base64, ...'
     *         ['url'] => string(37) 'https://domain.tld:5678/some-page.html'
     *         ['tags'] => array(4) {
     *             [0] => string(1) 'a'
     *             [1] => string(4) 'list'
     *             [2] => string(2) 'of'
     *             [3] => string(4) 'tags'
     *         }
     *         ['description'] => string(29) 'Some comments about this link'
     *         ['dateCreated'] => int(1234567890)
     *         ['public'] => bool(true)
     *     }
     *     [1]=> array(7) {
     *         // ...
     *     }
     * }
     *
     * @param string $data    The encoded string containing Netscape bookmarks
     * @param array  $context An optional set of options for the decoder; see below
     *
     * The $context array is a simple key=>value array, with the following supported keys:
     *
     * keep_nested_tags: boolean
     *     Tag links with parent folder names
     *
     * normalize_dates: boolean
     *     Whether parsed dates are expected to fall within a given date/time interval
     *
     * date_range: string
     *     Delta used to compute the "acceptable" date/time interval
     *
     * @return array An associative array containing parsed links
     *
     * @throws \Exception
     */
    public function decode(string $data, array $context = []): array
    {
        $keepNestedTags = $context[self::KEEP_NESTED_TAGS] ?? $this->defaultContext[self::KEEP_NESTED_TAGS];
        $normalizeDates = $context[self::NORMALIZE_DATES] ?? $this->defaultContext[self::NORMALIZE_DATES];
        $dateRange = $context[self::DATE_RANGE] ?? $this->defaultContext[self::DATE_RANGE];

        $items = [];
        $folderTags = [];
        $groupedFolderTags = [];

        $lines = explode("\n", $this->sanitizeString($data));

        foreach ($lines as $lineNumber => $line) {
            $item = [
                'name'        => null,
                'image'       => null,
                'url'         => null,
                'tags'        => null,
                'description' => null,
                'dateCreated' => null,
                'public'      => null,
            ];

            $this->logger->info('PARSING LINE #' . $lineNumber);
            $this->logger->debug('[#' . $lineNumber . '] Content: ' . $line);
            if (preg_match('/^<h\d.*>(.*)<\/h\d>/i', $line, $header)) {
                // a header is matched:
                // - links may be grouped in a (sub-)folder
                // - append the header's content to the folder tags
                $tag = static::sanitizeTags($header[1]);

                $groupedFolderTags[] = $tag;
                $folderTags = static::flattenTagsList($groupedFolderTags);
                $this->logger->debug('[#' . $lineNumber . '] Header found: ' . implode(' ', $tag));

                continue;
            } elseif (preg_match('/^<\/DL>/i', $line)) {
                array_pop($groupedFolderTags);
                // </DL> matched: stop using header value
                $folderTags = static::flattenTagsList($groupedFolderTags);
                $this->logger->debug('[#' . $lineNumber . '] Header ended: ' . implode(' ', $tag ?? []));
                continue;
            }

            if (preg_match('/<a/i', $line)) {
                // match bookmark url
                if (preg_match('/href="(.*?)"/i', $line, $href)) {
                    $item['url'] = $href[1] ? $href[1] : null;
                }
                $this->logger->debug(
                    sprintf('[#%s] %s', $lineNumber, $item['url'] ? 'URL found: ' . $item['url'] : 'Empty URL')
                );

                // match bookmark icon
                if (preg_match('/icon="(.*?)"/i', $line, $icon)) {
                    $item['image'] = $icon[1] ? $icon[1] : null;
                }
                $this->logger->debug(
                    sprintf(
                        '[#%s] %s',
                        $lineNumber,
                        $item['image'] ? 'ICON found: ' . substr($item['image'], 0, 50) . ' ...' : 'Empty ICON'
                    )
                );

                // match bookmark title
                if (preg_match('/<a.*?[^br]>(.*?)<\/a>/i', $line, $title)) {
                    $item['name'] = $title[1] ? $title[1] : null;
                }
                $this->logger->debug(
                    sprintf('[#%s] %s', $lineNumber, $item['name'] ? 'TITLE found: ' . $item['name'] : 'Empty TITLE')
                );

                // match bookmark note
                if (preg_match('/(description|note)="(.*?)"/i', $line, $note)) {
                    $note = $note[2];
                } elseif (preg_match('/<dd>(.*?)$/i', $line, $note)) {
                    // match bookmark note (alternative syntax)
                    $note = str_replace('<br>', "\n", $note[1]);
                }
                $item['description'] = $note ? $note : null;
                // phpcs:disable Generic.Files.LineLength
                $this->logger->debug(
                    sprintf(
                        '[#%s] %s',
                        $lineNumber,
                        $item['description'] ? 'DESCRIPTION found: ' . substr($item['description'], 0, 50) . ' ...' : 'Empty DESCRIPTION'
                    )
                );
                // phpcs:enable

                $tags = [];
                // merge nested tags
                if ($keepNestedTags) {
                    $tags = array_merge($tags, $folderTags);
                }

                // match bookmark tags
                if (preg_match('/(tags?|labels?|folders?)="(.*?)"/i', $line, $labels)) {
                    $separator = false !== strpos($labels[2], ',') ? ',' : ' ';
                    $tags = array_merge(
                        $tags,
                        static::splitTagString($labels[2], $separator)
                    );
                }
                $item['tags'] = $tags;

                // match bookmark date created
                if (preg_match('/add_date="(.*?)"/i', $line, $addDate)) {
                    $time = $this->parseDate($addDate[1], $normalizeDates, $dateRange);
                    $item['dateCreated'] = $time ? $time : null;
                }

                // phpcs:disable Generic.Files.LineLength
                $this->logger->debug(
                    sprintf(
                        '[#%s] %s',
                        $lineNumber,
                        $item['dateCreated'] ? 'DATE found: ' . $item['dateCreated'] : 'Empty DATE'
                    )
                );
                // phpcs:enable

                // match bookmark visibility
                if (preg_match('/(public|published|pub)="(.*?)"/i', $line, $public)) {
                    // bookmark visibility is public
                    $item['public'] = $this->parseBoolean($public[2]);
                } elseif (preg_match('/(private|shared)="(.*?)"/i', $line, $private)) {
                    // bookmark visibility is private (inverse values)
                    $item['public'] = !$this->parseBoolean($private[2]);
                }
                $this->logger->debug(
                    sprintf('[#%s] %s', $lineNumber, $item['public'] ? 'public' : 'private')
                );

                $this->logger->info('File parsing ended');
                $items[] = $item;
            }
        }

        return $items;
    }

    /**
     * Parses a formatted date.
     *
     * @see http://php.net/manual/en/datetime.formats.compound.php
     * @see http://php.net/manual/en/function.strtotime.php
     *
     * @param string $date           Formatted date
     * @param bool   $normalizeDates Normalize dates true or false
     * @param string $dateRange      String representing date range
     *
     * @return int Unix timestamp corresponding to a successfully parsed date,
     *             else current date and time
     *
     * @throws \Exception
     */
    public function parseDate(string $date, bool $normalizeDates, string $dateRange): int
    {
        if (strtotime('@' . $date)) {
            // Unix timestamp
            if ($normalizeDates) {
                $date = $this->normalizeDate($date, $dateRange);
            }

            return strtotime('@' . $date);
        } elseif (strtotime($date)) {
            // attempt to parse a known compound date/time format
            return strtotime($date);
        }

        // current date & time
        return time();
    }

    /**
     * Normalizes a date by supposing it is comprised in a given range.
     *
     * Although most bookmarking services return dates formatted as a Unix epoch
     * (seconds elapsed since 1970-01-01 00:00:00) or human-readable strings,
     * some services return microtime epochs (microseconds elapsed since
     * 1970-01-01 00:00:00.000000) WITHOUT using a delimiter for the microseconds
     * part...
     *
     * This is likely to raise issues in the distant future!
     *
     * @see https://stackoverflow.com/questions/33691428/datetime-with-microseconds
     * @see https://stackoverflow.com/questions/23929145/how-to-test-if-a-given-time-stamp-is-in-seconds-or-milliseconds
     * @see https://stackoverflow.com/questions/539900/google-bookmark-export-date-format
     * @see https://www.wired.com/2010/11/1110mars-climate-observer-report/
     *
     * @param string $epoch     Unix timestamp to normalize
     * @param string $dateRange String representing date range
     *
     * @return int Unix timestamp in seconds, within the expected range
     *
     * @throws \Exception
     */
    public function normalizeDate(string $epoch, string $dateRange): int
    {
        $date = new \DateTime('@' . $epoch);
        $maxDate = new \DateTime('+' . $dateRange);

        for ($i = 1; $date > $maxDate; ++$i) {
            // trim the provided date until it falls within the expected range
            $date = new \DateTime('@' . substr($epoch, 0, \strlen($epoch) - $i));
        }

        return $date->getTimestamp();
    }

    /**
     * Parses the value of a supposedly boolean attribute.
     *
     * @param string $value Attribute value to evaluate
     *
     * @return bool 'true' when the value is evaluated as true 'false' when the value is evaluated as false
     *              or the value is undefined
     */
    public function parseBoolean(string $value): bool
    {
        if (preg_match('/^(' . self::TRUE_PATTERN . ')$/i', $value)) {
            return true;
        }

        if (preg_match('/^(' . self::FALSE_PATTERN . ')$/i', $value)) {
            return false;
        }

        return false;
    }

    /**
     * Sanitizes the content of a string containing Netscape bookmarks.
     *
     * This removes:
     * - comment blocks
     * - metadata: DOCTYPE, H1, META, TITLE
     * - extra newlines, trailing spaces and tabs
     *
     * @param string $bookmark Original bookmark string
     *
     * @return string Sanitized bookmark string
     */
    public static function sanitizeString(string $bookmark): string
    {
        // remove comments
        $bookmark = preg_replace('@<!--.*?-->@mis', '', $bookmark);

        // keep one XML element per line to prepare for linear parsing
        $bookmark = preg_replace('@>(\s*?)<@mis', ">\n<", $bookmark);

        // remove unused metadata
        $bookmark = preg_replace('@(<!DOCTYPE|<META|<TITLE|<H1|<P).*\n@i', '', $bookmark);

        // trim whitespace
        $bookmark = trim($bookmark);

        // remove carriage returns
        $bookmark = str_replace("\r", '', $bookmark);

        // convert multiline descriptions to one-line descriptions
        // convert new line to <br>
        $bookmark = preg_replace_callback(
            '@<DD>(.*?)(</?(:?DT|DD|DL))@mis',
            function ($match) {
                return '<DD>' . str_replace("\n", '<br>', trim($match[1])) . PHP_EOL . $match[2];
            },
            $bookmark
        );

        // convert multiline descriptions inside <A> tags to one-line descriptions
        // convert new line to <br>
        $bookmark = preg_replace_callback(
            '@<A(.*?)</A>@mis',
            function ($match) {
                return '<A ' . str_replace("\n", '<br>', trim($match[1])) . '</A>';
            },
            $bookmark
        );

        // concatenate all data related to the same entry on the same line
        // e.g. <A HREF="...">My Link</A><DD>List<br>- item1<br>- item2
        $bookmark = preg_replace('@\n<br>@mis', '<br>', $bookmark);
        $bookmark = preg_replace('@\n<DD@i', '<DD', $bookmark);

        return $bookmark;
    }

    /**
     * Split tag string using provided separator.
     *
     * @param string $tagString Tag string
     * @param string $separator tag separator
     *
     * @return array List of tags (trimmed and filtered)
     */
    public static function splitTagString(string $tagString, string $separator): array
    {
        $tags = explode($separator, strtolower($tagString));

        // remove multiple consecutive whitespaces
        $tags = preg_replace('/\s{2,}/', ' ', $tags);

        return array_values(array_filter(array_map('trim', $tags)));
    }

    /**
     * Sanitizes a space-separated list of tags.
     *
     * This removes:
     * - duplicate whitespace
     * - leading punctuation
     * - undesired characters
     *
     * @param string $tagString Space-separated list of tags
     *
     * @return array List of sanitized tags
     */
    public static function sanitizeTags(string $tagString): array
    {
        $separator = false !== strpos($tagString, ',') ? ',' : ' ';
        $tags = static::splitTagString($tagString, $separator);

        foreach ($tags as $key => &$value) {
            if (ctype_alnum($value)) {
                continue;
            }

            $keepWhiteSpaces = ' ' !== $separator;

            $value = strtolower($value);

            // trim leading punctuation
            $value = preg_replace('/^[[:punct:]]/', '', $value);

            // trim all but alphanumeric characters, underscores and non-leading dashes
            $value = preg_replace('/[^\p{L}\p{N}\-_' . ($keepWhiteSpaces ? ' ' : '') . ']++/u', '', $value);

            if ('' == $value) {
                unset($tags[$key]);
            }
        }

        return array_values($tags);
    }

    /**
     * Flatten a multi-dimensions array of tags into a one-dimension array.
     *
     * @param array $groupedTags Array of arrays of tags
     *
     * @return array Flatten tags list
     */
    public static function flattenTagsList(array $groupedTags): array
    {
        return array_reduce(
            $groupedTags,
            function (array $carry, array $item) {
                return array_merge($carry, $item);
            },
            []
        );
    }
}
